File: CS/Scaffolding.DetailCollections/DepartmentContextDataModel/Runtime/DepartmentContextUnitOfWork.cs
using System;
using System.Linq;
using System.Data;
using System.Data.Entity;
using System.Linq.Expressions;
using System.Collections.Generic;
using Scaffolding.DetailCollections.Common.Utils;
using Scaffolding.DetailCollections.Common.DataModel;
using Scaffolding.DetailCollections.Common.DataModel.EntityFramework;
using Scaffolding.DetailCollections.Model;

namespace Scaffolding.DetailCollections.DepartmentContextDataModel {
    public class DepartmentContextUnitOfWork : DbUnitOfWork<DepartmentContext>, IDepartmentContextUnitOfWork {
        Lazy<ICourseRepository> coursesRepository;
        Lazy<IDepartmentRepository> departmentsRepository;
        Lazy<IEmployeeRepository> employeesRepository;

        public DepartmentContextUnitOfWork(DepartmentContext context)
            : base(context) {
            coursesRepository = new Lazy<ICourseRepository>(() => new CourseRepository(this));
            departmentsRepository = new Lazy<IDepartmentRepository>(() => new DepartmentRepository(this));
            employeesRepository = new Lazy<IEmployeeRepository>(() => new EmployeeRepository(this));
        }
        bool IDepartmentContextUnitOfWork.HasChanges() {
            return Context.ChangeTracker.HasChanges();
        }
        ICourseRepository IDepartmentContextUnitOfWork.Courses {
            get { return coursesRepository.Value; }
        }
        IDepartmentRepository IDepartmentContextUnitOfWork.Departments {
            get { return departmentsRepository.Value; }
        }
        IEmployeeRepository IDepartmentContextUnitOfWork.Employees {
            get { return employeesRepository.Value; }
        }
    }
}

File: VB/Scaffolding.DetailCollections/DepartmentContextDataModel/Runtime/DepartmentContextUnitOfWork.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Linq
Imports System.Data
Imports System.Data.Entity
Imports System.Linq.Expressions
Imports System.Collections.Generic
Imports Scaffolding.DetailCollections.Common.Utils
Imports Scaffolding.DetailCollections.Common.DataModel
Imports Scaffolding.DetailCollections.Common.DataModel.EntityFramework
Imports Scaffolding.DetailCollections.Model

Namespace Scaffolding.DetailCollections.DepartmentContextDataModel
	Public Class DepartmentContextUnitOfWork
		Inherits DbUnitOfWork(Of DepartmentContext)
		Implements IDepartmentContextUnitOfWork
		Private coursesRepository As Lazy(Of ICourseRepository)
		Private departmentsRepository As Lazy(Of IDepartmentRepository)
		Private employeesRepository As Lazy(Of IEmployeeRepository)

		Public Sub New(ByVal context As DepartmentContext)
			MyBase.New(context)
			coursesRepository = New Lazy(Of ICourseRepository)(Function() New CourseRepository(Me))
			departmentsRepository = New Lazy(Of IDepartmentRepository)(Function() New DepartmentRepository(Me))
			employeesRepository = New Lazy(Of IEmployeeRepository)(Function() New EmployeeRepository(Me))
		End Sub
		Private Function HasChanges() As Boolean Implements IDepartmentContextUnitOfWork.HasChanges
			Return Context.ChangeTracker.HasChanges()
		End Function
		Private ReadOnly Property Courses() As ICourseRepository Implements IDepartmentContextUnitOfWork.Courses
			Get
				Return coursesRepository.Value
			End Get
		End Property
		Private ReadOnly Property Departments() As IDepartmentRepository Implements IDepartmentContextUnitOfWork.Departments
			Get
				Return departmentsRepository.Value
			End Get
		End Property
		Private ReadOnly Property Employees() As IEmployeeRepository Implements IDepartmentContextUnitOfWork.Employees
			Get
				Return employeesRepository.Value
			End Get
		End Property
	End Class
End Namespace
